Add power beam sampling

mwilensky768 opened the pull request:
This adds a lot of changes to beam_example.py and related code to make it so that power beam sampling can be done.

Additionally, it adds some compatibility for more up-to-date pyuvdata and matvis versions. I also think we should switch to numpy>2 -- the only relevant update is the np.random API as far as I understand.

@philbull I'm requesting your review for vis_simulator.py

In particular, I've done some changes to follow pyuvdata's UVBeam/AnalyticBeam/BeamInterface stuff and I'm not sure if I made the relevant update everywhere in the code (in other words I fixed functions that I used during beam sampling but other simulator functions might need to follow up and I think you have a better knowledge of where these lines exist).

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+from matplotlib.colors import LogNorm, SymLogNorm
+from matplotlib.gridspec import GridSpec
+
+def do_vis_sim(args, ftime, times, freqs, ant_pos, Nants, ra, dec,
```
Perhaps `run_vis_sim` is more in-keeping with elsewhere in the codebase

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+        _sim_vis = np.load(sim_outpath)
+    return _sim_vis
+
+def get_pert_beam(
```
Perhaps just `perturbed_beam`.

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+    
+    Parameters:
+        args (Namespace):
+            Arguments that were parsed with argparse.
```
Best to be explicit about which ones are needed... This is the documentation after all!

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+        sin_pert_coeffs=np.zeros(8, dtype=float)
+):
+    """
+    Make a perturbed beam.
```
Add "A file called XXX will be saved in output_dir"

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+    outfile = f"{output_dir}/perturbed_beam_beamvals_seed_{seed}.npy"
+    load = os.path.exists(outfile)
+    save = not load
+    if seed is not None: # Ignore inputs and generate them randomly here
```
This behaviour should be mentioned in the docstring

philbull commented on beam_example.py:
```diff
@@ -1,28 +1,256 @@
+        nmax=args.nmax,
+        sqrt=args.per_ant, Nfeeds=2, 
+        num_modes_comp=args.Nbasis, 
+        save=save, cSL=args.csl,
```
Should be one kwarg per line

philbull commented on beam_example.py:
```diff
@@ -149,38 +390,28 @@
+    # Convert to radians
     array_lat = np.deg2rad(args.array_lat)
 
+    if "Vivaldi" in args.beam_file:
```
Case sensitive?

philbull commented on beam_example.py:
```diff
@@ -208,227 +439,874 @@
+                    ref_beam = UVBeam.from_file(args.beam_file)
+                    ref_beam.peak_normalize()
+            elif args.beam_type in ["gaussian", "airy"]:
+                # Underillimunated HERA dishes
```
Typo

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+        os.path.join(output_dir, "mmodes.npy"),
+        mmodes
+    )
+    per_source_Dmatr_out = os.path.join(output_dir, "Dmatr.npy")
```
I'd suggest adding a few more inline comments in this code block to say what each step is calculating

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+        beam_coeffs = np.zeros([Nants, args.Nfreqs, args.Nbasis, 1, 1])
+        if not args.perts_only:
+            if args.decent_prior:
+                # FIXME: Hardcode!, start at answer!
```
Note the FIXME here!

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+        if not args.perts_only:
+            if args.decent_prior:
+                # FIXME: Hardcode!, start at answer!
+                beam_coeffs += np.load("data/14m_airy_bessel_soln.npy")[None, None, :, None, None] 
```
This path should perhaps be defined at the top of the script to make it easier for people to find and modify, or be set using a command line argument (where the current path is the default)

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
-            bbeam = rhs_unflatten.flatten()
+        sig_freq = 0.1 * (freqs[-1] - freqs[0])
+        # FIXME: Hardcode!
+        cov_file = "data/ramped_variance_bessel_cov.npy" if args.decent_prior else None
```
Same as above for this filename

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+                
+                axlen = np.prod(shape)
+
+                # fPbpQBcCF->fbQcFBpPC
```
For future sanity, I would suggest giving (or referring to) a key that tells us what each index letter is

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+                    os.remove(file)
+                np.save(os.path.join(output_dir, f"beam_roundup_{np1}"), sample_arr)
+    else:
+        pow_beam_Dmatr_outfile = os.path.join(output_dir, "pow_beam_Dmatr.npy")
```
Please add a few more inline comments in the block below

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+            bess_sky_contraction = np.load(bsc_outpath)
+        else:
+            t0 = time.time()
+            bess_sky_contraction = hydra.beam_sampler.get_bess_sky_contraction(Dmatr_outer, 
```
Consider using kwargs instead of positional args. It makes errors less likely, and if we later decide to change these functions, it will save some head scratching.

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
-            beam_coeffs[:, :, ant_samp_ind] = 1.0 * x_soln_swap[:, :, :, :, 0] \
-                                                + 1.j * x_soln_swap[:, :, :, :, 1]
+                # Construct RHS vector
+                rhs_unflatten = hydra.beam_sampler.construct_rhs(data_use,
```
Similar comment as above regarding kwargs

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+        sparse_dmatr_recon = sparse_bmatr[:, None] * sparse_tmatr[None, :]
+
+        # fb,zab->fza but without einsum as the middleman
+        MAP_beam = np.tensordot(MAP_soln,
```
Also low on inline comments below

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+        ax[0, 0].set_title("MAP Beam")
+        fig.colorbar(im, ax=ax[0,0])
+
+        image_var = np.einsum("bB,azb,azB->az",
```
Ideally would have a comment explaining this operation

philbull commented on beam_example.py:
```diff
@@ -208,227 +439,874 @@
+
+            
+
+        PPD_Dmatr = pow_beam_Dmatr[:, 1::2, triu_inds[0], triu_inds[1]]
```
Needs comment

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+
+        PPD_Dmatr = pow_beam_Dmatr[:, 1::2, triu_inds[0], triu_inds[1]]
+
+        postdicted_mean = np.einsum(
```
Definitely needs comment

philbull commented on power_beam_example.py:
```diff
@@ -208,227 +439,874 @@
+            linthresh=1e-4
+        )
+
+
```
Overall comment on this file: Perhaps we should rename it. I know I started this bad habit of naming the _actual run script_ as `example.py`, but someone opening this 1307-line script and expecting to see a basic example would presumably fall off their chair :)

This script in itself is a substantial component of the software, so we may as well treat it as more than just an example? (Perhaps we can put some of the function defs in a utility .py file also, just to reduce the size of this file.)

mwilensky768 replied:
Yes I think some very mild refactoring might actually help here.

After reading through this again (and taking a break from it for a while), I think most of the power beam leg of the example should actually live in a `power_beam_sampler.py` with `construct_LHS` functions etc. In other words there's a lot of magic that is actually essential to incorporating the power beam sampler as a head of `hydra` that I've just left as mostly unlabled wizardry in this script. 

The current `beam_sampler.py` can then be `per_ant_beam_sampler.py` or something. The example can then be split in two, and we can think about what to name them. Sounds good?

philbull commented on hydra/sparse_beam.py:
```diff
@@ -62,9 +62,63 @@ def __init__(
+            Nfeeds (int):
+                Number of feeds to read in. This does not usually need to be set,
+                but may need to be set depending on pyuvdata version.
+            do_fit (bool):
```
Perhaps `fit_beam`?

philbull commented on hydra/sparse_beam.py:
```diff
@@ -62,9 +62,63 @@ def __init__(
+                Sparse fit coefficients from previous run if do_fit is False.
+            perturb (bool):
+                Whether to perturb the beam.
+            za_ml (float):
```
Needs units. Perhaps a more transparent name, e.g. `mainlobe_za`?

philbull commented on hydra/sparse_beam.py:
```diff
@@ -62,9 +62,63 @@ def __init__(
+                Parameter controlling the maximum strength of the sidelobe
+                perturbations.
+            gam (float):
+                Parameter controlling additional mainlobe perturbations (Not)
```
Why (Not)?

mwilensky768 replied:
The parentheses were supposed to go around the rest of this keyword's entry. I've fixed it now. However, this keyword's name sucks. I named it after the Greek variable in Samir's paper. OTOH, it is just like `csl` in how it's named, and it is kind of annoying to describe it physically. I also expect it will just not be used so maybe it's not worth the headache. LMK your thoughts.

philbull commented on hydra/sparse_beam.py:
```diff
@@ -62,9 +62,63 @@ def __init__(
+                A parameter determining the size in zenith angle for 
+                mainlobe-specific perturbations, not including coordinate
+                transformations.
+            dza (float):
```
Units

philbull commented on hydra/sparse_beam.py:
```diff
@@ -373,17 +431,23 @@ def sparse_fit_loop(
         Parameters:
-            fit_coeffs (array, complex):
-                Precomputed fit coefficients (if just evaluating).
             do_fit (bool):
```
See above.

mwilensky768 replied:
The user is supplying these and the unit is variable. I will instead put unit descriptions in the Returns section.

philbull commented on hydra/vis_simulator.py:
```diff
@@ -320,6 +298,49 @@ def vis_sim_per_source(
     # Return visibilities with or without multiple polarization channels
     return vis if polarized else vis[0, 0]
 
+def get_interp_beam_for_sim(
```
Needs docstring

philbull commented on hydra/vis_simulator.py:
```diff
@@ -320,6 +298,49 @@ def vis_sim_per_source(
+    else:
+        # Here we have already asserted that the beam is a power beam and
+        # has only one polarization, so we just evaluate that one.
+        if spw_axis_present: # TODO: This branch is deprecated if we just mandate up-to-date pyuvdata
```
Let's mandate that. We can, as we don't have any users yet!

philbull commented on requirements.txt:
```diff
@@ -1,9 +1,8 @@
 matvis
 pygdsm
-healpy
 pyuvsim
```
Let's require the updated version, as annoying as that is

philbull reviewed (commented):
Sorry, I got most of the way through this review and then went on leave. Here is a partial review -- mostly just minor stuff about naming conventions or docstrings.

mwilensky768 commented:
Hi @philbull just to update you: I resolved a few comments in beam_example.py but have quite a few more to go for the rest of the review.

mwilensky768 commented:
HI @philbull I have basically addressed everything that won't be addressed in a refactor of `beam_example.py`. Waiting to see hear your thoughts about splitting that into two files as per my previous comment, since it is a small but nontrivial amount of work. Please reach out if you have questions.

mwilensky768 commented:
Hi @philbull I just refactored beam_example into four new files: `per_antenna_beam_example.py`, `power_beam_example.py`, `power_beam_sampler.py`, and `beam_example_utils.py`. I think this last file probably has useful helper functions for setting up simulations in general, so it may not have been a purely aesthetic endeavor. I tested that this recovers the beam paper II plots. I also fixed some merge conflicts related to pyuvdata compatibility. 

I think this is ready to merge.
